Index: src/java/org/nanocontainer/reflection/DefaultSoftCompositionPicoContainer.java
===================================================================
RCS file: /scm/picocontainer/java/nanocontainer/src/java/org/nanocontainer/reflection/DefaultSoftCompositionPicoContainer.java,v
retrieving revision 1.3
diff -u -r1.3 DefaultSoftCompositionPicoContainer.java
--- src/java/org/nanocontainer/reflection/DefaultSoftCompositionPicoContainer.java	25 Jun 2004 12:03:24 -0000	1.3
+++ src/java/org/nanocontainer/reflection/DefaultSoftCompositionPicoContainer.java	26 Jun 2004 22:19:00 -0000
@@ -57,8 +57,12 @@
         this(new DefaultComponentAdapterFactory(), parent);
     }
 
+    public DefaultSoftCompositionPicoContainer(ComponentAdapterFactory caf) {
+        this(caf, null);
+    }
+
     public DefaultSoftCompositionPicoContainer() {
-        this(null);
+        this(new DefaultComponentAdapterFactory(), null);
     }
 
     public Object getComponentInstance(Object componentKey) {
Index: src/test/org/nanocontainer/script/groovy/NanoGroovyBuilderTestCase.groovy
===================================================================
RCS file: /scm/picocontainer/java/nanocontainer/src/test/org/nanocontainer/script/groovy/NanoGroovyBuilderTestCase.groovy,v
retrieving revision 1.7
diff -u -r1.7 NanoGroovyBuilderTestCase.groovy
--- src/test/org/nanocontainer/script/groovy/NanoGroovyBuilderTestCase.groovy	25 Jun 2004 13:40:15 -0000	1.7
+++ src/test/org/nanocontainer/script/groovy/NanoGroovyBuilderTestCase.groovy	26 Jun 2004 22:19:01 -0000
@@ -94,23 +94,20 @@
         }
     }
 
-//    TODO NANO-90
-//    TODO - The CAF is not being passed into the ctor of teh SoftPico.
-//    See NanoGroovyBuilder.java
-//    void testInstantiateWithBespokeComponentAdaptor() {
+    void testInstantiateWithBespokeComponentAdaptor() {
 
-//        sb = new StringBuffer();
+        sb = new StringBuffer();
 
-//        builder = new NanoGroovyBuilder()
-//        pico = builder.container(adapterFactory:new TestComponentAdapterFactory(sb)) {
-//            component(key:WebServerConfig, class:DefaultWebServerConfig)
-//            component(key:WebServer, class:WebServerImpl)
-//        }
+        builder = new NanoGroovyBuilder()
+        pico = builder.container(adapterFactory:new TestComponentAdapterFactory(sb)) {
+            component(key:WebServerConfig, class:DefaultWebServerConfig)
+            component(key:WebServer, class:WebServerImpl)
+        }
 
-//        startAndStop(pico)
+        startAndStop(pico)
 
-//        assertTrue(sb.toString().indexOf("called") != -1)
-//    }
+        assertTrue(sb.toString().indexOf("called") != -1)
+    }
 
     void testInstantiateWithInlineConfiguration() {
 
